README.md
# ChatGPT extension for VSCode

[![Github stars](https://img.shields.io/github/stars/yaleh/chatgpt-vscode)](https://github.com/yaleh/chatgpt-vscode)

ChatGPT AI is a Visual Studio Code extension that enables you to automatically write new code, ask questions, find bugs, refactor code, and more using the official OpenAI API or ChatGPT (via the [unofficial ChatGPT API](https://github.com/transitive-bullshit/chatgpt-api)). The extension provides a chat interface that you can use to interact with the AI model and get suggestions and explanations for your code.

Supercharge your coding with AI-powered assistance! Automatically write new code from scratch, ask questions, get explanations, refactor code, find bugs and more 🚀✨

## 03/01/2023: Introducing the official ChatGTP API! Now supported with the powerful gpt-3.5-turbo-0301 model.

### Links:

- **[Github Repository](https://github.com/yaleh/chatgpt-vscode)**
- **[Visual Studio Code Marketplace](https://marketplace.visualstudio.com/items?itemName=YaleHuang.chatgpt-ai)**

<br>

<img src="examples/main.png" alt="Refactoring selected code using chatGPT"/>

## Features
- 💡 **Ask general questions** or use code snippets from the editor to query ChatGPT via an input box in the sidebar
- 🖱️ Right click on a code selection and run one of the context menu **shortcuts**
	- automatically write documentation for your code
	- explain the selected code
	- refactor or optimize it
	- find problems with it
	- complete the code according to the comments
- 💻 View ChatGPT's responses in a panel next to the editor
- 🚀 See the response as it is being generated **in real time**
- 💬 Ask **follow-up questions** to the response (conversation context is maintained)
- 📝 **Insert code snippets** from the AI's response into the active editor or **copy code snippets** by clicking on them

## Requirements
ChatGPT AI requires Visual Studio Code version 1.73.0 or later to be installed on your system.

## Installation

To use this extension, install it from the VSCode marketplace.

1. After the installation is complete, you will need to add your OpenAI API key to the extension settings in VSCode. To do this, open the `Settings` panel by going to the `File` menu and selecting `Preferences`, then `Settings`.
2. In the search bar, type `ChatGPT` to filter the settings list.
3. In the ChatGPT section, enter your API key in the top field

After completing these steps, the extension should be ready to use.

## API Modes

Currently, there are two modes for accessing the API: "ChatGPTAPI" and "ChatGPTUnofficialProxyAPI".
### ChatGPTAPI

To use this extension, you will need an `API key` from OpenAI. To obtain one, follow these steps:

1. Go to [OpenAI's website](https://platform.openai.com/account/api-keys). If you don't have an account, you will need to create one or sign up using your Google or Microsoft account.
2. Click on the `Create new secret key` button.
3. Copy the key and paste it into the `API Key` field in the extension settings.

### ChatGPTUnofficialProxyAPI

To use the ChatGPTUnofficialProxyAPI, you need an OpenAI access token from the ChatGPT web app. You can obtain the access token in either of two ways:

* Use acheong08/OpenAIAuth, which is a Python script to automatically log in and get an access token. This method works only with email and password accounts; it does not support accounts where you authenticate via Microsoft or Google.
* Manually get an access token by logging in to the ChatGPT web app and opening https://chat.openai.com/api/auth/session, which will return a JSON object containing your access token string.

Access tokens last for approximately eight hours.

## Building from source (not applicable for VSCode marketplace version)

To build the extension from source, clone the repository and run `npm install` to install the dependencies. You have to change some code in `chatgpt` module because VSCode runtime does not support `fetch`. Open `node_modules/chatgpt/dist/index.js` and add the following code at the top of the file:

```js
import fetch from 'node-fetch'
```

Then remove the following lines (around line 15):

```js
// src/fetch.ts
var fetch = globalThis.fetch;
if (typeof fetch !== "function") {
  throw new Error("Invalid environment: global fetch not defined");
}
```

You also need to copy `encoder.json` and `vocab.bpe` from `node_modules/gpt-3-encoder/` into `dist/` folder. You can do this by running the following commands:

```bash
cp node_modules/gpt-3-encoder/{encoder.json,vocab.bpe} dist/
```

## Using the Extension

To use the extension, open a text editor in Visual Studio Code and open the ChatGPT panel by clicking on the ChatGPT icon in the sidebar. This will open a panel with an input field where you can enter your prompt or question. By clicking enter, it will be sent to ChatGPT. Its response will be displayed below the input field in the sidebar (note that it may take some time for it to be calculated).

<img src="examples/create.png" alt="Writing new code using chatGPT" width="500"/>

You can also select a code snippet in the editor and then enter a prompt in the side panel, or right-click and select "Ask ChatGPT". The **selected code will be automatically appended** to your query when it is sent to the AI. This can be useful for generating code snippets or getting explanations for specific pieces of code.

<img src="examples/explain.png" alt="Refactoring selected code using chatGPT"/>

To **insert a code snippet** from the AI's response into the editor, simply click on the code block in the panel. The code will be automatically inserted at the cursor position in the active editor.

<img src="examples/refactor.png" alt="chatGPT explaining selected code"/>

You can select some code in the editor, right click on it and choose one of the following **shortcuts** from the context menu:
#### Commands

The following commands are available:

* Ask ChatGPT: This command will provide a prompt for you to enter any query.
* ChatGPT: Explain selection: This command will explain what the selected code does.
* ChatGPT: Refactor selection: This command will attempt to refactor the selected code.
* ChatGPT: Find problems: This command will look for problems/errors in the selected code, fix them, and provide an explanation.
* ChatGPT: Optimize selection: This command will attempt to optimize the selected code.
* ChatGPT: Complete code: This command will complete the code according to the last comments of the selection.

`Ask ChatGPT` is also available when nothing is selected. For the other four commands, you can **customize the exact prompt** that will be sent to the AI by editing the extension settings in VSCode Preferences.


Because ChatGPT is a conversational AI, you can ask follow-up questions to the response. The conversation context is maintained between queries, so you can ask multiple questions in a row. 
To **reset the conversation context**, click `ctrl+shift+p` and select `ChatGPT: Reset Conversation`.

---

Please note that this extension is currently a proof of concept and may have some limitations or bugs. We welcome feedback and contributions to improve the extension. Also check out [CodeGPT](https://github.com/timkmecl/codegpt) extension that uses official OpenAI API and also supports other GPT3 models.

## License

ChatGPT AI is licensed under the MIT License. See the LICENSE file for more information.

## Credits

- This wouldn't be possible without OpenAI's [ChatGPT](https://chat.openai.com/chat)
- The extension makes use of [chatgpt-api](https://github.com/transitive-bullshit/chatgpt-api) (by [Travis Fischer](https://github.com/transitive-bullshit)), which uses unofficial ChatGPT API in order to login and communicate with it.
- The project was started by [mpociot](https://github.com/mpociot/) and [timkmecl](https://github.com/timkmecl)
- `v0.3` inspired by [barnesoir/chatgpt-vscode-plugin](https://github.com/barnesoir/chatgpt-vscode-plugin) and [gencay/vscode-chatgpt](https://github.com/gencay/vscode-chatgpt)

CHANGELOG.md
# Change Log

All notable changes to the "chatgpt" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

## [0.6.4] - 2023-02-25

### Added
* Applied new version of the ChatGPT module.
### Changed
* Applied new styles for markdown rendering using marked and highlight.js.
### Fixed
* Fixed minor bugs and improved performance.

## [0.6.3] - 2022-12-31
# Added

* Initial release of the "chatgpt" extension.
*Added functionality to chat with an AI-powered chatbot using OpenAI's GPT-3 API.

media/main.js
// @ts-ignore 

// This script will be run within the webview itself
// It cannot access the main VS Code APIs directly.
(function () {
  const vscode = acquireVsCodeApi();

  let response = '';

  // Handle messages sent from the extension to the webview
  window.addEventListener("message", (event) => {
    const message = event.data;
    switch (message.type) {
      case "addResponse": {
        response = message.value;
        setResponse();
        break;
      }
      case "clearResponse": {
        response = '';
        break;
      }
      case "setPrompt": {
        document.getElementById("prompt-input").value = message.value;
        break;
      }
      case "setDisplayingMode": {
        mode = message.value;
        if (mode === "asking") {
          toggleStopButton(true);
        } else {
          toggleStopButton(false);
        }
        break;
      }
    }
  });

  function fixCodeBlocks(response) {
    // Use a regular expression to find all occurrences of the substring in the string
    const REGEX_CODEBLOCK = new RegExp('\`\`\`', 'g');
    const matches = response.match(REGEX_CODEBLOCK);

    // Return the number of occurrences of the substring in the response, check if even
    const count = matches ? matches.length : 0;
    if (count % 2 === 0) {
      return response;
    } else {
      // else append ``` to the end to make the last code block complete
      return response.concat('\n\`\`\`');
    }

  }

  let lastMessageId = null;

  function setResponse(messageId = null) {
    marked.setOptions({
      renderer: new marked.Renderer(),
      highlight: function (code, lang) {
        return hljs.highlightAuto(code).value;
      },
      langPrefix: 'hljs language-', // highlight.js css expects a top-level 'hljs' class.
      pedantic: false,
      gfm: true,
      breaks: false,
      sanitize: false,
      smartypants: false,
      xhtml: false
    });

    response = fixCodeBlocks(response);
    html = marked.parse(response);

    var responseDiv = document.getElementById("responses");

    if ((responseDiv.childElementCount > 0 && responseDiv.lastChild !== null) && (messageId === null || messageId === lastMessageId)) {
      // Update the existing response
      responseDiv.lastChild.innerHTML = html;
    } else {
      // Create a new div and append it to the "response" div
      var newDiv = document.createElement('div');
      newDiv.classList.add('response');
      newDiv.innerHTML = html;
      responseDiv.appendChild(newDiv);
    }

    var codeBlocks = document.querySelectorAll('pre > code');
    for (var i = 0; i < codeBlocks.length; i++) {
      // Check if innertext starts with "Copy code"
      if (codeBlocks[i].innerText.startsWith("Copy code")) {
        codeBlocks[i].innerText = codeBlocks[i].innerText.replace("Copy code", "");
      }

      codeBlocks[i].classList.add("inline-flex", "max-w-full", "overflow-hidden", "rounded-sm", "cursor-pointer");

      var insertButton = document.createElement('button');
      insertButton.textContent = "Insert";
      insertButton.classList.add("text-xs", "font-medium", "leading-5", "text-white", "bg-indigo-600", "hover:bg-indigo-500", "focus:outline-none", "focus:ring", "focus:ring-indigo-500", "focus:ring-opacity-50", "px-2", "py-1", "rounded-sm");

      var codeBlock = codeBlocks[i];
      codeBlock.parentNode.insertBefore(insertButton, codeBlock);

      insertButton.style.display = "block";
      codeBlock.style.display = "block";

      insertButton.addEventListener('click', function (e) {
        e.preventDefault();
        var code = this.nextElementSibling.innerText;
        vscode.postMessage({
          type: 'codeSelected',
          value: code
        });
      });

      const d = document.createElement('div');
      d.innerHTML = codeBlocks[i].innerHTML;
      codeBlocks[i].innerHTML = null;
      codeBlocks[i].appendChild(d);
      d.classList.add("code");
    }
  }

  function toggleStopButton(enableld) {
    const button = document.getElementById('stop-button');
    if (enableld) {
      button.disabled = false;
      button.classList.remove('bg-gray-400', 'cursor-not-allowed');
      button.classList.add('bg-red-600', 'hover:bg-red-700');
    } else {
      button.disabled = true;
      button.classList.remove('bg-red-600', 'hover:bg-red-700');
      button.classList.add('bg-gray-400', 'cursor-not-allowed');
    }
  }

  toggleStopButton(false);

  // Listen for keyup events on the prompt input element
  document.getElementById('prompt-input').addEventListener('keyup', function (e) {
    // If the key that was pressed was the Enter key
    if (e.keyCode === 13) {
      vscode.postMessage({
        type: 'prompt',
        value: this.value
      });
    }
  });

  // Listen for click events on the stop button
  document.getElementById('stop-button').addEventListener('click', function () {
    vscode.postMessage({
      type: 'abort'
    });
  });
})();


src/extension.ts
import * as vscode from 'vscode';
import { ChatGPTAPI, ChatGPTUnofficialProxyAPI } from 'chatgpt';

import * as path from 'path';
import * as fs from 'fs';
import * as cheerio from 'cheerio';

type AuthInfo = {
	mode?: string,
	apiKey?: string,
	accessToken?: string,
	proxyUrl?: string
};
type Settings = {selectedInsideCodeblock?: boolean, keepConversation?: boolean, timeoutLength?: number};
type WorkingState = 'idle' | 'asking';

export function activate(context: vscode.ExtensionContext) {

	console.log('activating extension "chatgpt"');
	// Get the settings from the extension's configuration
	const config = vscode.workspace.getConfiguration('chatgpt');

	// Create a new ChatGPTViewProvider instance and register it with the extension's context
	const provider = new ChatGPTViewProvider(context.extensionPath, context.extensionUri);

	// Put configuration settings into the provider
	provider.setAuthenticationInfo({
		mode: config.get('mode'),
		apiKey: config.get('apiKey'),
		accessToken: config.get('accessToken'),
		proxyUrl: config.get('proxyUrl') === "Custom" ? config.get('customProxyUrl') : config.get('proxyUrl')
	});
	provider.setSettings({
		selectedInsideCodeblock: config.get('selectedInsideCodeblock') || false,
		keepConversation: config.get('keepConversation') || false,
		timeoutLength: config.get('timeoutLength') || 60,
	});

	// Register the provider with the extension's context
	context.subscriptions.push(
		vscode.window.registerWebviewViewProvider(ChatGPTViewProvider.viewType, provider,  {
			webviewOptions: { retainContextWhenHidden: true }
		})
	);


	const commandHandler = (command:string) => {
		const config = vscode.workspace.getConfiguration('chatgpt');
		const prompt = config.get(command) as string;
		provider.searchSelection(prompt);
	};

	// Register the commands that can be called from the extension's package.json
	context.subscriptions.push(
		vscode.commands.registerCommand('chatgpt.ask', () =>
			vscode.window.showInputBox({ prompt: 'What do you want to do?' })
				.then((value) => {
					if (value !== undefined && value !== null) {
						provider.searchSelection(value);
					}
				})
		),
		vscode.commands.registerCommand('chatgpt.explain', () => commandHandler('promptPrefix.explain')),
		vscode.commands.registerCommand('chatgpt.refactor', () => commandHandler('promptPrefix.refactor')),
		vscode.commands.registerCommand('chatgpt.optimize', () => commandHandler('promptPrefix.optimize')),
		vscode.commands.registerCommand('chatgpt.findProblems', () => commandHandler('promptPrefix.findProblems')),
		vscode.commands.registerCommand('chatgpt.documentation', () => commandHandler('promptPrefix.documentation')),
		vscode.commands.registerCommand('chatgpt.complete', () => commandHandler('promptPrefix.complete')),
		vscode.commands.registerCommand('chatgpt.resetConversation', () => provider.resetConversation())
	);

	// Change the extension's session token or settings when configuration is changed
	vscode.workspace.onDidChangeConfiguration((event: vscode.ConfigurationChangeEvent) => {
		if (
			event.affectsConfiguration('chatgpt.mode') ||
			event.affectsConfiguration('chatgpt.apiKey') ||
			event.affectsConfiguration('chatgpt.accessToken') ||
			event.affectsConfiguration('chatgpt.proxyUrl')
		) {
			const config = vscode.workspace.getConfiguration('chatgpt');
			provider.setAuthenticationInfo({
				mode: config.get('mode'),
				apiKey: config.get('apiKey'),
				accessToken: config.get('accessToken'),
				proxyUrl: config.get('proxyUrl') === "Custom" ? config.get('customProxyUrl') : config.get('proxyUrl')
			});

			// clear conversation
			provider.resetConversation();
		} else if (event.affectsConfiguration('chatgpt.selectedInsideCodeblock')) {
			const config = vscode.workspace.getConfiguration('chatgpt');
			provider.setSettings({ selectedInsideCodeblock: config.get('selectedInsideCodeblock') || false });
		} else if (event.affectsConfiguration('chatgpt.keepConversation')) {
			const config = vscode.workspace.getConfiguration('chatgpt');
			provider.setSettings({ keepConversation: config.get('keepConversation') || false });
		} else if (event.affectsConfiguration('chatgpt.timeoutLength')) {
			const config = vscode.workspace.getConfiguration('chatgpt');
			provider.setSettings({ timeoutLength: config.get('timeoutLength') || 60 });
		}
	});
}





class ChatGPTViewProvider implements vscode.WebviewViewProvider {
	public static readonly viewType = 'chatgpt.chatView';
	private _view?: vscode.WebviewView;

	private _chatGPTAPI?: ChatGPTAPI | ChatGPTUnofficialProxyAPI;
	private _conversation?: any;

	// An AbortController for _chatGPTAPI
	private _abortController = new AbortController();

	private _response?: string;
	private _prompt?: string;
	private _fullPrompt?: string;
	private _currentMessageNumber = 0;

	private _workingState: WorkingState;

	private _settings: Settings = {
		selectedInsideCodeblock: false,
		keepConversation: true,
		timeoutLength: 60
	};
	private _authInfo?: AuthInfo;

	// In the constructor, we store the URI of the extension
	constructor(private readonly _extensionPath: string, private readonly _extensionUri: vscode.Uri) {
		this._workingState = 'idle';
	}
	
	// Set the API key and create a new API instance based on this key
	public setAuthenticationInfo(authInfo: AuthInfo) {
		this._authInfo = authInfo;
		this._newAPI();
	}

	public setSettings(settings: Settings) {
		this._settings = {...this._settings, ...settings};
	}

	public getSettings() {
		return this._settings;
	}

	private _setWorkingState(mode: WorkingState) {
		this._workingState = mode;
		this._view?.webview.postMessage({ type: 'setWorkingState', value: this._workingState});
	}

	// This private method initializes a new ChatGPTAPI instance
	private _newAPI() {
		console.log("New API");

		this._conversation = null;
		this._currentMessageNumber = 0;

		if (!this._authInfo) {
			console.warn("Invalid auth info, please set working mode and related auth info.");
			return;
		}
		if (this._authInfo?.mode === "ChatGPTAPI") {
			if (!this._authInfo?.apiKey) {
				console.warn("API key not set, please go to extension settings (read README.md for more info)");
			}else{
				this._chatGPTAPI = new ChatGPTAPI({
					apiKey: this._authInfo.apiKey,
					debug: false
				});
			}
		} else if (this._authInfo?.mode === "ChatGPTUnofficialProxyAPI") {
			if (!this._authInfo?.accessToken) {
				console.warn("Access token not set, please go to extension settings (read README.md for more info)");
			}else if (!this._authInfo?.proxyUrl) {
				console.warn("Proxy URL not set, please go to extension settings (read README.md for more info)");
			}else{
				this._chatGPTAPI = new ChatGPTUnofficialProxyAPI({
					accessToken: this._authInfo.accessToken,
					apiReverseProxyUrl: this._authInfo.proxyUrl,
					debug: false
				});
			}			
		}

	}

	public resolveWebviewView(
		webviewView: vscode.WebviewView,
		context: vscode.WebviewViewResolveContext,
		_token: vscode.CancellationToken,
	) {
		this._view = webviewView;

		// set options for the webview, allow scripts
		webviewView.webview.options = {
			enableScripts: true,
			localResourceRoots: [
				this._extensionUri
			]
		};

		// set the HTML for the webview
		webviewView.webview.html = this._getHtmlForWebview(webviewView.webview);

		// add an event listener for messages received by the webview
		webviewView.webview.onDidReceiveMessage(data => {
			switch (data.type) {
				case 'webviewLoaded':
				{
					this._view?.webview.postMessage({ type: 'setWorkingState', value: this._workingState});
					break;
				}
				case 'codeSelected':
					{
						let code = data.value;
						const snippet = new vscode.SnippetString();
						snippet.appendText(code);
						// insert the code as a snippet into the active text editor
						vscode.window.activeTextEditor?.insertSnippet(snippet);
						break;
					}
				case 'prompt':
					{
						this.searchSelection(data.value);
						break;
					}
				case 'abort':
					{
						this.abort();
						break;
					}
			}
		});
	}


	public async resetConversation() {
		console.log(this, this._conversation);
		if (this._conversation) {
			this._conversation = null;
		}
		this._currentMessageNumber = 0;
		this._prompt = '';
		this._response = '';
		this._fullPrompt = '';
		this._view?.webview.postMessage({ type: 'setPrompt', value: '' });
		this._view?.webview.postMessage({ type: 'addResponse', value: '' });
	}


	public async searchSelection(prompt?:string) {
		this._prompt = prompt;
		if (!prompt) {
			prompt = '';
		};

		// Check if the ChatGPTAPI instance is defined
		if (!this._chatGPTAPI) {
			this._newAPI();
		}

		// focus gpt activity from activity bar
		if (!this._view) {
			await vscode.commands.executeCommand('chatgpt.chatView.focus');
		} else {
			this._view?.show?.(true);
		}
		
		let response = '';
		this._response = '';
		// Get the selected text of the active editor
		const selection = vscode.window.activeTextEditor?.selection;
		const selectedText = vscode.window.activeTextEditor?.document.getText(selection);
		let searchPrompt = '';

		if (selection && selectedText) {
			// If there is a selection, add the prompt and the selected text to the search prompt
			if (this._settings.selectedInsideCodeblock) {
				searchPrompt = `${prompt}\n\`\`\`\n${selectedText}\n\`\`\``;
			} else {
				searchPrompt = `${prompt}\n${selectedText}\n`;
			}
		} else {
			// Otherwise, just use the prompt if user typed it
			searchPrompt = prompt;
		}
		this._fullPrompt = searchPrompt;

		this.sendMessageAndGetResponse(searchPrompt);
	}

	private async sendMessageAndGetResponse(searchPrompt: string): Promise<string> {
		let response = '';
		if (!this._chatGPTAPI) {
			response = '[ERROR] "API key not set or wrong, please go to extension settings to set it (read README.md for more info)"';
		} else {
			// If successfully signed in
			console.log("sendMessage");

			// Make sure the prompt is shown
			this._view?.webview.postMessage({ type: 'setPrompt', value: this._prompt });

			// Increment the message number
			this._currentMessageNumber++;

			this._setWorkingState('asking');

			try {
				// Send the search prompt to the ChatGPTAPI instance and store the response
				let currentMessageNumber = this._currentMessageNumber;
				const res = await this._chatGPTAPI.sendMessage(searchPrompt, {
					onProgress: (partialResponse) => {
						// If the message number has changed, don't show the partial response
						if (this._currentMessageNumber !== currentMessageNumber) {
							return;
						}
						console.log("onProgress");
						if (this._view && this._view.visible) {
							response = partialResponse.text;
							this._view.webview.postMessage({ type: 'addResponse', value: partialResponse.text });
						}
					},
					timeoutMs: (this._settings.timeoutLength || 60) * 1000,
					abortSignal: this._abortController.signal,
					...this._conversation
				});

				this._view?.webview.postMessage({ type: 'setWorkingState', value: 'idle'});

				if (this._currentMessageNumber !== currentMessageNumber) {
					return '';
				}

				response = res.text;
				if (this._settings.keepConversation){
					this._conversation = {
						conversationId: res.conversationId,
						parentMessageId: res.id
					};
				}
			} catch (e) {
				this._view?.webview.postMessage({ type: 'setWorkingState', value: 'idle'});

				console.error(e);
				response += `\n\n---\n[ERROR] ${e}`;
			}
		}

		// Saves the response
		this._response = response;

		// Show the view and send a message to the webview with the response
		if (this._view) {
			this._view.show?.(true);
			this._view.webview.postMessage({ type: 'addResponse', value: response });
		}

		return response;
	}

	public abort(){
		this._view?.webview.postMessage({ type: 'setWorkingState', value: 'idle'});

		this._abortController?.abort();
		// reset the controller
		this._abortController = new AbortController();
	}

	private _getHtmlForWebview(webview: vscode.Webview) {
		const indexHtmlPath = path.join(this._extensionPath, 'media', 'html', 'index.html');
		const indexHtml = fs.readFileSync(indexHtmlPath, 'utf8');
		
		const $ = cheerio.load(indexHtml);
		$('#responses').empty();
		
		const scriptUri = webview.asWebviewUri((vscode.Uri as any).joinPath(this._extensionUri, 'dist', 'main.js'));
		const tailwindUri = webview.asWebviewUri((vscode.Uri as any).joinPath(this._extensionUri, 'media', 'scripts', 'tailwind.min.js'));
		const highlightcssUri = webview.asWebviewUri((vscode.Uri as any).joinPath(this._extensionUri, 'media', 'styles', 'highlight-vscode.min.css'));

		return $.html()
			.replace('{{tailwindUri}}', tailwindUri.toString())
			.replace('{{highlightcssUri}}', highlightcssUri.toString())
			.replace('{{scriptUri}}', scriptUri.toString());
	}
}

// This method is called when your extension is deactivated
export function deactivate() {}

src/webview/main.ts
// @ts-ignore 

import * as marked from 'marked';
import hljs from 'highlight.js';

declare const acquireVsCodeApi: () => any;

interface ChatResponse {
  id: string;
  text: string;
  parentMessageId?: string;
  conversationId?: string;
}

interface ChatRequest {
  id?: string;
  text: string;
  parentMessageId?: string;
  conversationId?: string;
}

interface ChatEvent {
  text: string;
}

// This script will be run within the webview itself
// It cannot access the main VS Code APIs directly.
(function () {
  const vscode = acquireVsCodeApi();

  let response = '';
  let workingState = 'idle';

  // Handle messages sent from the extension to the webview
  window.addEventListener("message", (event: MessageEvent) => {
    const message = event.data;
    switch (message.type) {
      case "addResponse": {
        // const chatResponse: ChatResponse = message.value;
        // response = message.value;
        updateResponse(message.value as ChatResponse);
        break;
      }
      case "addReqeust": {
        updateRequest(message.values as ChatRequest);
        break;
      }
      case "addEvent": {
        updateEvent(message.values as ChatEvent);
        break;
      }
      case "clearResponses": {
        clearResponses();
        break;
      }
      case "setPrompt": {
        const promptInput = document.getElementById("prompt-input") as HTMLInputElement;
        promptInput.value = message.value;
        break;
      }
      case "setWorkingState": {
        setWorkingState(message.value);
        break;
      }
      case "setConversationId": {
        updateConversationId(message.value);
        break;
      }
    }
  });

  function updateConversationId(id: string) {
    const conversationId = id || "/";
    const conversationIdText = `Conversation ID: ${conversationId}`;
    const conversationIdElement = document.getElementById("conversation-id") as HTMLDivElement;
    conversationIdElement.innerText = conversationIdText;
  }

  function fixCodeBlocks(response: string) {
    const REGEX_CODEBLOCK = new RegExp('\`\`\`', 'g');
    const matches = response.match(REGEX_CODEBLOCK);

    const count = matches ? matches.length : 0;
    if (count % 2 === 0) {
      return response;
    } else {
      return response.concat('\n\`\`\`');
    }
  }

  let lastResponse: ChatResponse | null = null;

  function updateResponse(response: ChatResponse){
    const responsesDiv = document.getElementById("responses") as HTMLDivElement;
    let updatedResponseDiv: HTMLElement | null = null;

    if (responsesDiv.childElementCount > 0 && responsesDiv.lastChild !== null && (response.id === null || response?.id === lastResponse?.id)) {
      // Update the existing response
      updatedResponseDiv = responsesDiv.lastChild as HTMLElement;
    } else {
      // Create a new div and append it to the "response" div
      const newDiv = document.createElement('div');
      newDiv.classList.add('response');
      responsesDiv.appendChild(newDiv);
      updatedResponseDiv = newDiv;
    }

    updateResponseDiv(updatedResponseDiv, response);
    lastResponse = response;
  }

  function updateResponseDiv(div: HTMLElement, response: ChatResponse) {
    const markedOptions: marked.MarkedOptions = {
      renderer: new marked.Renderer(),
      highlight: (code: string, lang: string) => {
        return hljs.highlightAuto(code).value;
      },
      langPrefix: 'hljs language-',
      pedantic: false,
      gfm: true,
      breaks: false,
      sanitize: false,
      smartypants: false,
      xhtml: false
    };

    marked.setOptions(markedOptions);

    var fixedResponseText = fixCodeBlocks(response.text);
    const html = marked.parse(fixedResponseText);

    div.innerHTML = html;

    const codeBlocks = div.querySelectorAll('pre > code');

    for (let i = 0; i < codeBlocks.length; i++) {
      const codeBlock = codeBlocks[i] as HTMLElement;
      const innerText = codeBlock.innerText;
    
      const insertButton = createCodeSnippetButton('Insert', 'bg-indigo-600', (e) => {
        e.preventDefault();
        const code = codeBlock?.innerText;
        if (code) {
          vscode.postMessage({
            type: 'codeSelected',
            value: code
          });
        }
      });
    
      const copyButton = createCodeSnippetButton('Copy', 'bg-blue-400', (e) => {
        e.preventDefault();
        const code = codeBlock.innerText;
        navigator.clipboard.writeText(code).then(() => {
          console.log('Code copied to clipboard');
          const popup = createCodeSnippetPopup('Code copied to clipboard');
          document.body.appendChild(popup);
          setTimeout(() => {
            popup.remove();
          }, 2000);
        });
      });
    
      codeBlock.parentNode?.parentNode?.insertBefore(insertButton, codeBlock.parentNode);
      codeBlock.parentNode?.parentNode?.insertBefore(copyButton, codeBlock.parentNode);
    
      codeBlock.classList.add('hljs');
    }

  }

  function createCodeSnippetButton(text: string, color: string, clickHandler: (e: MouseEvent) => void) {
    const button = document.createElement('button');
    button.textContent = text;
    button.classList.add(
      'text-xs',
      'font-medium',
      'leading-5',
      'text-white',
      'hover:bg-gray-500',
      'focus:outline-none',
      'focus:ring',
      'focus:ring-opacity-50',
      'px-2',
      'py-1',
      'rounded-sm',
      color
    );
    button.addEventListener('click', clickHandler);
    return button;
  }
  
  function createCodeSnippetPopup(text: string) {
    const popup = document.createElement('div');
    popup.textContent = text;
    popup.classList.add(
      'text-xs',
      'font-medium',
      'leading-5',
      'text-white',
      'bg-green-500',
      'p-2',
      'rounded-sm',
      'absolute',
      'top-0',
      'right-0',
      'mt-2',
      'mr-2'
    );
    return popup;
  }

  function clearResponses(){
    const responsesDiv = document.getElementById("responses") as HTMLDivElement;
    // delete all children of responsesDiv
    while (responsesDiv.firstChild) {
      responsesDiv.removeChild(responsesDiv.firstChild);
    }
    lastResponse = null;
  }

  function updateRequest(request: ChatRequest){
    // TODO
  }

  function updateEvent(event: ChatEvent){
    // TODO
  }

  function setWorkingState(state: string) {
    workingState = state;
    toggleStopButton(workingState === 'asking');
    const workingStateElement = document.getElementById('working-state') as HTMLElement;
    if (workingState === 'asking') {
      workingStateElement.innerText = 'Thinking...';
    } else {
      workingStateElement.innerText = '';
    }
  }

  function toggleStopButton(enabled: boolean) {
    const button = document.getElementById('stop-button') as HTMLButtonElement;
    if (enabled) {
      button.disabled = false;
      button.classList.remove('bg-gray-400', 'cursor-not-allowed');
      button.classList.add('bg-red-600', 'hover:bg-red-700');
    } else {
      button.disabled = true;
      button.classList.remove('bg-red-600', 'hover:bg-red-700');
      button.classList.add('bg-gray-400', 'cursor-not-allowed');
    }
  }

  // Listen for keyup events on the prompt input element
  const promptInput = document.getElementById('prompt-input') as HTMLInputElement;
  promptInput.addEventListener('keyup', (e: KeyboardEvent) => {
    // If the key combination that was pressed was Ctrl+Enter
    if (e.keyCode === 13 && e.ctrlKey) {
      sendMessage(promptInput.value);
    }
  });

  // Listen for click events on the "send-request" button
  const sendButton = document.getElementById('send-request') as HTMLButtonElement;
  sendButton.addEventListener('click', () => {
    sendMessage(promptInput.value);
  });

  // Function to send a message to the extension
  function sendMessage(value: string) {
    vscode.postMessage({
      type: 'prompt',
      value: value
    });
  }

  // Listen for click events on the stop button
  const stopButton = document.getElementById('stop-button') as HTMLButtonElement;
  stopButton.addEventListener('click', () => {
    vscode.postMessage({
      type: 'abort'
    });
  });

  // Listen for click events on the reset button and send message resetConversation
  const resetButton = document.getElementById('reset-button') as HTMLButtonElement;
  resetButton.addEventListener('click', () => {
    vscode.postMessage({
      type: 'resetConversation'
    });
  });

  vscode.postMessage({type: 'webviewLoaded'});
})();

